src/lib.rs

```diff
@@ -85,51 +99,64 @@ impl FsPool {
         &self,
         path: P,
         opts: WriteOptions,
-    ) -> FsWriteSink {
+    ) -> Result<FsWriteSink<E>, Error> {
         ::write::new(self, path, opts)
     }
 
     /// Returns a `Sink` to send bytes to be written to the supplied file.
-    pub fn write_file(
-        &self,
-        file: fs::File,
-    ) -> FsWriteSink {
+    pub fn write_file(&self, file: fs::File) -> FsWriteSink<E> {
         ::write::new_from_file(self, file)
     }
 
     /// Returns a `Future` that resolves when the target file is deleted.
-    pub fn delete<P: AsRef<Path> + Send + 'static>(&self, path: P) -> FsFuture<()> {
-        fs(self.cpu_pool.spawn_fn(move || fs::remove_file(path)))
+    pub fn delete<P: AsRef<Path> + Send + 'static>(&self, path: P) -> Result<FsFuture<()>, Error> {
+        let (tx, rx) = oneshot::channel();
```

Comment:
The futures crate already has this pattern available, with [`oneshot::spawn_fn`](https://docs.rs/futures/0.1.21/futures/sync/oneshot/fn.spawn_fn.html) (and `oneshot::spawn`). That would probably simplify things for us!

Reply:
I'm not going to be able to address this concern, since the `Executor` trait requires that you define the kind of future you're going to execute on it. Right now, I have FsPool containing a

```rust
Box<Executor<Box<Future<Item = (), Error = ()> + Send>>>
```

Which is an Executor trait object that can execute Future trait objects. This is a problem with `spawn_fn`, which wants to execute a Future, not a Future Trait Object. The executor type for our needs would be

```rust
Box<Executor<Execute<Lazy<{some closure}, Result<{closure success}, {closure error}>>>>>
```
which isn't possible to do, especially if we want to be able to execute more than one closure.

Furthermore, it seems futures 0.2.1 does not have a concept of `spawn_fn` for oneshot. Maybe this was moved somewhere else, but a quick search coudln't find it.

Reply:
Futures 0.2.1's executor makes this better by assuming an executor will execute a 
```rust
Box<Future<Item = (), Error = Never> + Send + 'static>
```
rather than letting the implementer specify the kind of future to be executed, so the type of future an executor executes is not tied to the type that implements executor.

Reply:
If you can think of a way around this issue, I'd love to hear about it.

Reply:
Even trying to use spawn over spawn_fn runs into issues, since now we need the executor type to be
```rust
Box<Executor<Execute<Box<Future<Item = SomeItem, Error = io::Error>>>>>
```
which again ties our FsPool to the Item type of the future.

Looking at the source code for `oneshot::spawn`, it seems like it does exactly what I implemented, but panics if there are errors.

Reply:
Ah, yea, you're totally right, my bad. It's fine that it's implemented here, I just thought maybe we didn't need to repeat what already existed. No problem...